### PR TITLE
Deploy Gaia-X Authority did document

### LINKED ISSUE #151: GAIA-X Authority - DID
Feature #78 
After #18 

## Description

MVD GAIA-X authority should have DID document related to it.
Similar story for the reference: https://github.com/agera-edc/MinimumViableDataspace/issues/44

## Acceptance Criteria

- [ ] Well-known DID endpoint for MVD GAIA-X authority deployed.
- [ ] DID document contains public key

## Tasks
- [ ] Deploy DID document for GAIA-X in github workflow

### PR DESCRIPTION
**Pull request by Izzzu**
Deployment of GAIA-X Authority DID in Deploy pipeline.
DID document contains public key in the same format as Registration Service did document. 

### Further notes

- renaming the `registry` to either `authority` for did document resources and `registrationservice` for Registration Service resources for consistency
- removing `branches: [ main ]` from CD workflow to enable checks in all PRs

linked to #151

### CODE REVIEW

File: .github/workflows/deploy.yaml [hunk 1]
```diff
@@ -140,11 +140,17 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      - name: 'Generate key'
+      - name: 'Generate Registration Service key'
         run: |
-          openssl ecparam -name prime256v1 -genkey -noout -out key.pem
-          openssl ec -in key.pem -pubout -out key.public.pem
-          docker run -i danedmunds/pem-to-jwk:1.2.1 --public --pretty < key.public.pem > key.public.jwk
+          openssl ecparam -name prime256v1 -genkey -noout -out registrykey.pem
```
Comment (algattik) [inline .github/workflows/deploy.yaml]: could we extract an action?

File: .github/workflows/deploy.yaml [hunk 2]
```diff
@@ -140,11 +140,17 @@ jobs:
...
+          openssl ec -in registrykey.pem -pubout -out registrykey.public.pem
+          docker run -i danedmunds/pem-to-jwk:1.2.1 --public --pretty < registrykey.public.pem > registrykey.public.jwk
+
+      - name: 'Generate GAIA-X Authority key'
```
Comment (algattik) [inline .github/workflows/deploy.yaml:143]: should logically be first

File: .github/workflows/deploy.yaml [hunk 3]
```diff
@@ -140,11 +140,17 @@ jobs:
...
       - uses: actions/checkout@v2
 
-      - name: 'Generate key'
+      - name: 'Generate Registration Service key'
         run: |
-          openssl ecparam -name prime256v1 -genkey -noout -out key.pem
-          openssl ec -in key.pem -pubout -out key.public.pem
-          docker run -i danedmunds/pem-to-jwk:1.2.1 --public --pretty < key.public.pem > key.public.jwk
+          openssl ecparam -name prime256v1 -genkey -noout -out registrykey.pem
+          openssl ec -in registrykey.pem -pubout -out registrykey.public.pem
```
Comment (algattik) [inline .github/workflows/deploy.yaml]: "registration service" should not be misnamed "registry"

File: .github/workflows/deploy.yaml [hunk 4]
```diff
@@ -140,11 +140,17 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      - name: 'Generate key'
+      - name: 'Generate Registration Service key'
```
Comment (algattik) [inline .github/workflows/deploy.yaml]: this is the Dataspace Authority key, not directly related to Registration Service

File: deployment/terraform/dataspace/variables.tf [hunk 5]
```diff
@@ -35,12 +35,22 @@ variable "application_sp_object_id" {
...
-variable "key_file" {
-  description = "name of a file containing the private key in PEM format"
+variable "key_file_registry" {
+  description = "name of a file containing the Registration Service private key in PEM format"
   default     = null
 }
 
-variable "public_key_jwk_file" {
-  description = "name of a file containing the public key in JWK format"
+variable "key_file_gaiax" {
```
Comment (algattik) [inline deployment/terraform/dataspace/variables.tf]: what is this for?

**Review by algattik — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy.yaml (hunk 3 above)
Comment (chrislomonico) [reply .github/workflows/deploy.yaml]: a few places in the files, the use of `registry` vs `registration service` is used; should we be mindful of these as well for the sake of consistency?

File: .github/workflows/deploy.yaml (hunk 2 above)
Comment (Izzzu) [reply .github/workflows/deploy.yaml:143]: done

File: .github/workflows/deploy.yaml (hunk 1 above)
Comment (Izzzu) [reply .github/workflows/deploy.yaml]: done

File: .github/workflows/deploy.yaml (hunk 3 above)
Comment (Izzzu) [reply .github/workflows/deploy.yaml]: Yes, good point to keep the consistency in mind. I changed the name of the key to registrationservice. We still have 'registry' in some places (eg terraform) but I wouldn't make it a part of this PR to rename everything.

File: .github/workflows/deploy.yaml (hunk 4 above)
Comment (Izzzu) [reply .github/workflows/deploy.yaml]: agree, renamed

File: deployment/terraform/dataspace/variables.tf (hunk 5 above)
Comment (Izzzu) [reply deployment/terraform/dataspace/variables.tf]: not needed, removed :)

File: deployment/terraform/dataspace/variables.tf [hunk 6]
```diff
@@ -35,12 +35,17 @@ variable "application_sp_object_id" {
   description = "object id of application's service principal object"
 }
 
-variable "key_file" {
-  description = "name of a file containing the private key in PEM format"
+variable "key_file_registry" {
```
Comment (chrislomonico) [inline deployment/terraform/dataspace/variables.tf]: nit: minor but relates to the topic of `registry` vs `registration service`; ultimately this is up to you to decide

File: deployment/terraform/dataspace/variables.tf [hunk 7]
```diff
@@ -35,12 +35,17 @@ variable "application_sp_object_id" {
...
+  description = "name of a file containing the Registration Service private key in PEM format"
   default     = null
 }
 
-variable "public_key_jwk_file" {
-  description = "name of a file containing the public key in JWK format"
+variable "public_key_jwk_file_registry" {
```
Comment (chrislomonico) [inline deployment/terraform/dataspace/variables.tf]: nit: minor but relates to the topic of `registry` vs `registration service`; ultimately this is up to you to decide (same as above)

File: .github/workflows/deploy.yaml (hunk 3 above)
Comment (algattik) [reply .github/workflows/deploy.yaml]: Created https://github.com/agera-edc/MinimumViableDataspace/issues/180

File: .github/workflows/cd.yaml [hunk 8]
```diff
@@ -2,7 +2,6 @@ name: CD
 
 on:
   pull_request:
-    branches: [ main ]
```
Comment (chrislomonico) [inline .github/workflows/cd.yaml:5]: Assuming that this removal is intention to ensure that this runs on PR's into other branches

**Review by chrislomonico — CHANGES_REQUESTED** [review]
Please review comments provided

**Review by algattik — CHANGES_REQUESTED** [review]
The DID and private key are related not just to "registration service", but really to "dataspace authority" which is more than that. Names should reflect this

File: deployment/terraform/dataspace/variables.tf (hunk 7 above)
Comment (Izzzu) [reply deployment/terraform/dataspace/variables.tf]: I renamed the `registry` to either `authority` for did document resources and 'registrationservice` for Registration Service resources

Comment (Izzzu) [reply deployment/terraform/dataspace/variables.tf]: I renamed the `registry` to either `authority` for did document resources and 'registrationservice` for Registration Service resources


> The DID and private key are related not just to "registration service", but really to "dataspace authority" which is more than that. Names should reflect this

indeed, I renamed the `registry` to either `authority` for did document resources and `registrationservice` for Registration Service resources

Comment (Izzzu) [reply deployment/terraform/dataspace/variables.tf]: I renamed the `registry` to either `authority` for did document resources and 'registrationservice` for Registration Service resources.

File: .github/workflows/deploy.yaml (hunk 3 above)
Comment (Izzzu) [reply .github/workflows/deploy.yaml]: After discussing with Alexandre offline
I renamed the `registry` to either `authority` for did document resources and `registrationservice` for Registration Service resources

**Comment by Izzzu** [comment]
> The DID and private key are related not just to "registration service", but really to "dataspace authority" which is more than that. Names should reflect this

indeed, I renamed the `registry` to either `authority` for did document resources and `registrationservice` for Registration Service resources

File: .github/actions/generate-key/action.yml [hunk 9]
```diff
@@ -0,0 +1,18 @@
+name: "Generate key"
+description: "Generate key"
```
Comment (algattik) [inline .github/actions/generate-key/action.yml]: could we extend this description

File: .github/actions/generate-key/action.yml [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+inputs:
+  keyname:
+    description: 'Name of the key file name'
```
Comment (algattik) [inline .github/actions/generate-key/action.yml]:
```suggestion
    description: 'Prefix for key file names'
```

File: .github/actions/generate-key/action.yml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+    required: true
+
+runs:
+  using: "composite"
+  steps:
+    - name: 'Generate GAIA-X Authority key'
```
Comment (algattik) [inline .github/actions/generate-key/action.yml]: needs generalisation

File: .github/actions/generate-key/action.yml [hunk 12]
```diff
@@ -0,0 +1,18 @@
+name: "Generate key"
+description: "Generate key"
+
+inputs:
+  keyname:
```
Comment (algattik) [inline .github/actions/generate-key/action.yml]:
```suggestion
  keyFileNamePrefix:
```

File: .github/workflows/deploy.yaml [hunk 13]
```diff
@@ -140,11 +140,15 @@ jobs:
...
-          docker run -i danedmunds/pem-to-jwk:1.2.1 --public --pretty < key.public.pem > key.public.jwk
+      - name: 'Generate GAIA-X Authority key'
+        uses: ./.github/actions/generate-key
+        with:
+          keyname: gaiaxkey
+
+      - name: 'Generate Dataspace Authority key'
+        uses: ./.github/actions/generate-key
+        with:
+          keyname: registrationservicekey
```
Comment (algattik) [inline .github/workflows/deploy.yaml]:
```suggestion
          keyname: authoritykey
```

File: .github/workflows/deploy.yaml [hunk 14]
```diff
@@ -241,11 +251,10 @@ jobs:
     steps:
       - uses: actions/checkout@v2
 
-      - name: 'Generate key'
-        run: |
-          openssl ecparam -name prime256v1 -genkey -noout -out key.pem
-          openssl ec -in key.pem -pubout -out key.public.pem
-          docker run -i danedmunds/pem-to-jwk:1.2.1 --public --pretty < key.public.pem > key.public.jwk
+      - name: 'Generate Participant's key'
```
Comment (algattik) [inline .github/workflows/deploy.yaml]:
```suggestion
      - name: 'Generate Participant key'
```

File: deployment/terraform/dataspace/main.tf [hunk 15]
```diff
@@ -133,39 +135,79 @@ resource "azurerm_storage_account" "did" {
...
       "https://www.w3.org/ns/did/v1",
       {
-        "@base" = local.did_url
+        "@base" = local.authority_did_url
       }
     ],
     "verificationMethod" = [
       {
-        "id"           = "#identity-key-1"
+        "id"           = "#identity-key-registration-service"
```
Comment (algattik) [inline deployment/terraform/dataspace/main.tf]:
```suggestion
        "id"           = "#identity-key-authority"
```

**Review by chrislomonico — APPROVED** [review]
open questions/comments addressed

File: deployment/terraform/dataspace/variables.tf [hunk 16]
```diff
@@ -35,12 +35,17 @@ variable "application_sp_object_id" {
   description = "object id of application's service principal object"
 }
 
-variable "key_file" {
-  description = "name of a file containing the private key in PEM format"
+variable "key_file_authority" {
+  description = "name of a file containing the Registration Service private key in PEM format"
```
Comment (cpeeyush) [inline deployment/terraform/dataspace/variables.tf:39]: nit: By reading description and the variable name it gives the impression that authority and registration service are related as its usage registration service private key as per the description.

No action needed on this PR but we can discuss it later.

File: deployment/terraform/dataspace/main.tf [hunk 17]
```diff
@@ -130,42 +132,82 @@ resource "azurerm_storage_account" "did" {
...
-resource "azurerm_key_vault_secret" "did_key" {
+resource "azurerm_key_vault_secret" "authority_did_key" {
   name = local.connector_name
   # Create did_key secret only if key_file value is provided. Default key_file value is null.
-  count        = var.key_file == null ? 0 : 1
-  value        = file(var.key_file)
-  key_vault_id = azurerm_key_vault.registry.id
+  count        = var.key_file_authority == null ? 0 : 1
+  value        = file(var.key_file_authority)
+  key_vault_id = azurerm_key_vault.registrationservice.id
```
Comment (cpeeyush) [inline deployment/terraform/dataspace/main.tf:140]: nit: Same as below comment, If registration service does not need a key vault, then maybe we do not deploy one for that and rename to `authority` as currently in this PR registration service key vault is used to store authority did private key only.

No action for now can be discussed in a follow-up PR.

**Review by cpeeyush — APPROVED** [review]
Approved with suggestions which can be discussed later, no action needed for this PR at the moment. Thank you.

**Pull request merged**